File: src/math/prng.ts
/**
 * Seeded PRNG using mulberry32 algorithm.
 * Deterministic: same seed always produces the same sequence.
 */
export class PRNG {
  private state: number

  constructor(seed: number) {
    this.state = seed >>> 0
  }

  /** Returns a float in [0, 1) */
  next(): number {
    this.state = (this.state + 0x6d2b79f5) >>> 0
    let t = this.state
    t = Math.imul(t ^ (t >>> 15), t | 1)
    t ^= t + Math.imul(t ^ (t >>> 7), t | 61)
    return ((t ^ (t >>> 14)) >>> 0) / 0x100000000
  }

  nextFloat(min: number, max: number): number {
    return min + this.next() * (max - min)
  }

  nextInt(min: number, max: number): number {
    return Math.floor(this.nextFloat(min, max + 1))
  }

  pick<T>(arr: readonly T[]): T {
    const item = arr[Math.floor(this.next() * arr.length)]
    if (item === undefined) throw new Error('pick called on empty array')
    return item
  }

  /** Box-Muller transform — returns a standard normal (mean=0, σ=1) sample */
  nextGaussian(): number {
    const u1 = this.next()
    const u2 = this.next()
    // Avoid log(0)
    const safe = u1 < 1e-10 ? 1e-10 : u1
    return Math.sqrt(-2 * Math.log(safe)) * Math.cos(2 * Math.PI * u2)
  }

  /** Gaussian sample clamped to [min, max] with given mean and standard deviation */
  nextGaussianClamped(mean: number, stddev: number, min: number, max: number): number {
    const raw = mean + this.nextGaussian() * stddev
    return Math.max(min, Math.min(max, raw))
  }
}

// ─── Wang hash for chunk seed derivation ────────────────────────────────────

function wangHash(n: number): number {
  n = ((n ^ 61) ^ (n >>> 16)) >>> 0
  n = Math.imul(n, 9) >>> 0
  n = (n ^ (n >>> 4)) >>> 0
  n = Math.imul(n, 0x27d4eb2d) >>> 0
  n = (n ^ (n >>> 15)) >>> 0
  return n
}

/**
 * Derives a deterministic seed for a specific chunk from the galaxy seed.
 * Two different (cx, cy) pairs always produce different seeds.
 */
export function chunkSeed(galaxySeed: number, cx: number, cy: number): number {
  // Map negative coords to unique positive integers before hashing
  const ux = cx < 0 ? -2 * cx - 1 : 2 * cx
  const uy = cy < 0 ? -2 * cy - 1 : 2 * cy
  let h = wangHash(galaxySeed >>> 0)
  h = wangHash(h ^ wangHash(ux >>> 0))
  h = wangHash(h ^ wangHash(uy >>> 0))
  return h >>> 0
}

File: generation/prng.py
"""
Seeded PRNG using mulberry32 algorithm.
Bit-exact port of src/math/prng.ts — uses & 0xFFFFFFFF instead of >>> 0.
"""
import math


def _imul(a: int, b: int) -> int:
    """32-bit integer multiplication (Math.imul equivalent)."""
    return ((a & 0xFFFFFFFF) * (b & 0xFFFFFFFF)) & 0xFFFFFFFF


class PRNG:
    def __init__(self, seed: int) -> None:
        self.state = seed & 0xFFFFFFFF

    def next(self) -> float:
        """Returns a float in [0, 1)."""
        self.state = (self.state + 0x6D2B79F5) & 0xFFFFFFFF
        t = self.state
        t = _imul(t ^ (t >> 15), t | 1)
        t = (t ^ (t + _imul(t ^ ((t >> 7) & 0xFFFFFFFF), t | 61))) & 0xFFFFFFFF
        return ((t ^ (t >> 14)) & 0xFFFFFFFF) / 0x100000000

    def next_float(self, min_val: float, max_val: float) -> float:
        return min_val + self.next() * (max_val - min_val)

    def next_int(self, min_val: int, max_val: int) -> int:
        return int(self.next_float(min_val, max_val + 1))

    def pick(self, arr: list):
        idx = int(self.next() * len(arr))
        return arr[idx]

    def next_gaussian(self) -> float:
        """Box-Muller transform — standard normal (mean=0, σ=1)."""
        u1 = self.next()
        u2 = self.next()
        safe = u1 if u1 >= 1e-10 else 1e-10
        return math.sqrt(-2 * math.log(safe)) * math.cos(2 * math.pi * u2)

    def next_gaussian_clamped(self, mean: float, stddev: float, min_val: float, max_val: float) -> float:
        raw = mean + self.next_gaussian() * stddev
        return max(min_val, min(max_val, raw))


# ── Wang hash for chunk seed derivation ──────────────────────────────────────

def wang_hash(n: int) -> int:
    n = ((n ^ 61) ^ (n >> 16)) & 0xFFFFFFFF
    n = (n * 9) & 0xFFFFFFFF
    n = (n ^ (n >> 4)) & 0xFFFFFFFF
    n = (n * 0x27D4EB2D) & 0xFFFFFFFF
    return (n ^ (n >> 15)) & 0xFFFFFFFF


def chunk_seed(galaxy_seed: int, cx: int, cy: int) -> int:
    """Derives a deterministic seed for a specific chunk from the galaxy seed."""
    ux = (-2 * cx - 1) if cx < 0 else (2 * cx)
    uy = (-2 * cy - 1) if cy < 0 else (2 * cy)
    h = wang_hash(galaxy_seed & 0xFFFFFFFF)
    h = wang_hash(h ^ wang_hash(ux & 0xFFFFFFFF))
    h = wang_hash(h ^ wang_hash(uy & 0xFFFFFFFF))
    return h
